### PR TITLE
add pobox to group

### PR DESCRIPTION
**Pull request by chereseeriepa**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -13,6 +13,9 @@ const bfe = require('ssb-bfe')
 
 const Method = require('./method')
 
+// TODO: remove when real keys are plugged in
```
Comment (chereseeriepa) [inline index.js:16]: TODO

File: index.js [hunk 2]
```diff
@@ -241,17 +244,26 @@ function init (ssb, config) {
 
           const { groupId: subgroupId, groupKey, groupInitMsg } = data
 
-          // TODO: generate a dmKey and attach to returned data
+          // TODO: remove this and add real code that generates the keys
```
Comment (chereseeriepa) [inline index.js:247]: TODO

File: method/group/add-pobox.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
+const Crut = require('ssb-crut')
+const PoBox = require('../../spec/group/add-pobox')
+
+module.exports = function GroupPobox (ssb, keystore, state) {
+  const poBox = new Crut(ssb, PoBox)
+
+  function isValidRoot (content) {
```
Comment (chereseeriepa) [inline method/group/add-pobox.js:7]: @mixmix I couldnt get spec.isRoot/spec.isUpdate to preserve the errorsString..... :( Any tips so i dont have to do it manually??

File: method/group/add-pobox.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+    if (!poBox.spec.isUpdate(T)) {
+      isValidUpdate.error = poBox.spec.isUpdate.errorsString
+      return false
+    }
+
+    return true
+  }
+
+  // TODO: choose whether to pass in the groupInit OR use the groupId to get it
```
Comment (chereseeriepa) [inline method/group/add-pobox.js:46]: TODO: 

Would like your opinion on this @mixmix

File: method/group/add-pobox.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        set: { publicKey: keys.publicKey, secretKey: keys.secretKey }
+      },
+      root,
+      tangles: {
+        poBox: {
+          root: groupInit,
+          previous: [groupInit] // TODO calculate previous for poBox tangle
+        },
+
+        group: { root: groupInit, previous: [root] } // TODO: need a way to validate this root
```
Comment (chereseeriepa) [inline method/group/add-pobox.js:63]: TODO:

@mixmix should i also validate this?

File: method/group/add-pobox.js (hunk 3 above)
Comment (chereseeriepa) [reply method/group/add-pobox.js:7]: NOTE: reasoning for these two functions

- couldnt use isRoot on the content which it didnt have `poBox: { root: null, previous: null }`, but i still wanted to make sure the staticProps were validated!!

Comment (chereseeriepa) [reply method/group/add-pobox.js:7]: LMK if this is okay, or theres a better way to write this!

File: method/group/add-pobox.js (hunk 4 above)
Comment (mixmix) [reply method/group/add-pobox.js:46]: you have the keystore, so I think you can use the groupId to get that info
would certainly tidy this up

Comment (mixmix) [reply method/group/add-pobox.js:46]: const { root } = keystore.group.get(groupId)

root is groupInitMsgId!

OH, and :fire: you don't have a guarentee that you have this groupId in the keystore. So do this:

```js
const info = keystore.group.get(groupId)
if (!info) return cb(new Error('unknown groupId: ' + groupId)
```

File: method/group/add-pobox.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  return function groupPobox (groupId, keys, groupInit, cb) {
+    const { root } = keystore.group.get(groupId)
+
+    const content = {
+      type: 'group/poBox',
+      version: 'v1',
+      keys: {
+        set: { publicKey: keys.publicKey, secretKey: keys.secretKey }
+      },
+      root,
```
Comment (mixmix) [inline method/group/add-pobox.js:56]: don't think we need this as have it in the tangle roots

File: method/group/add-pobox.js (hunk 3 above)
Comment (mixmix) [reply method/group/add-pobox.js:7]: DANG, I forgot about the staticProps if this is an update.
Technicaly I think we should just be using `isUpdate` if we're wanting to crut this later...

I reckon we can ditch the staticProps: root, version

File: spec/lib/field-types.js [hunk 7]
```diff
@@ -1,6 +1,22 @@
 const Overwrite = require('@tangle/overwrite')
+const isBase64 = require('is-canonical-base64')()
```
Comment (mixmix) [inline spec/lib/field-types.js:2]: :fire: I think json schema patterns need to be strings, but I could be wrong, and this is a regexp

in past I've done:

```js
pattern: isBase64.toString().slice(1, -1) // the slice cuts off the start/end '/'
```

Comment (mixmix) [reply spec/lib/field-types.js:2]: I suggest wring a simple test for this fieldType and seeing what that says

File: method/group/add-pobox.js (hunk 3 above)
Comment (chereseeriepa) [reply method/group/add-pobox.js:7]: Yay, that gets rid of all that code!

File: method/group/add-pobox.js (hunk 4 above)
Comment (chereseeriepa) [reply method/group/add-pobox.js:46]: Added!

File: spec/lib/field-types.js (hunk 7 above)
Comment (chereseeriepa) [reply spec/lib/field-types.js:2]: I think you're right, changed the value to a random one and it didnt fail, writing test now :)

Comment (chereseeriepa) [reply spec/lib/field-types.js:2]: I think you're right, changed the value to a random one and it didnt fail, writing test now :)

File: method/group/add-pobox.js [hunk 8]
```diff
@@ -4,71 +4,28 @@ const PoBox = require('../../spec/group/add-pobox')
...
-      return false
-    }
-
-    return true
-  }
-
-  // TODO: choose whether to pass in the groupInit OR use the groupId to get it
-  return function groupPobox (groupId, keys, groupInit, cb) {
-    const { root } = keystore.group.get(groupId)
+    const { root } = info
```
Comment (chereseeriepa) [inline method/group/add-pobox.js:12]: after looking at the keystore i realised that `root` is `groupInitMsg.key`

File: spec/lib/field-types.js [hunk 9]
```diff
@@ -6,16 +6,16 @@ const overwrite = (schema, opts = {}) => Overwrite({ valueSchema: schema, ...opt
...
     properties: {
       publicKey: {
-        type: ['string', 'null'],
+        type: ['string'],
         pattern: '^ssb://'
       },
-      privateKey: {
-        type: ['string', 'null'],
-        pattern: isBase64
+      secretKey: {
```
Comment (chereseeriepa) [inline spec/lib/field-types.js:16]: Took me ages to realise that i had a stupid typo here :/ was wondering why it kept passing even though it was wrong ugh

File: method/group/add-pobox.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    if (!info) return cb(new Error('unknown groupId: ' + groupId))
+
+    const { root } = info
+
+    const content = {
+      type: 'group/poBox',
+      keys: {
+        set: { publicKey, secretKey }
+      },
+      tangles: {
```
Comment (mixmix) [inline method/group/add-pobox.js:19]:
```js
/* dummy tangle - this get auto updated by publish hook */
```

File: method/group/add-pobox.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
+const Crut = require('ssb-crut')
+const PoBox = require('../../spec/group/add-pobox')
+
+module.exports = function GroupPobox (ssb, keystore, state) {
```
Comment (mixmix) [inline method/group/add-pobox.js:4]: GroupAddPoBox

File: method/group/add-pobox.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+      type: 'group/poBox',
+      keys: {
+        set: { publicKey, secretKey }
+      },
+      tangles: {
+        poBox: { root, previous: [root] }, // TODO calculate previous for poBox tangle
+        group: { root, previous: [root] } // TODO: need a way to validate this root
+        // NOTE: this is a dummy entry which is over-written in publish hook
+        // it's needed to pass isValid
+      },
```
Comment (mixmix) [inline method/group/add-pobox.js:24]: TODO - check this publish hook is working

File: method/group/add-pobox.js [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    const { root } = info
+
+    const content = {
+      type: 'group/poBox',
+      keys: {
+        set: { publicKey, secretKey }
+      },
+      tangles: {
+        poBox: { root, previous: [root] }, // TODO calculate previous for poBox tangle
+        group: { root, previous: [root] } // TODO: need a way to validate this root
```
Comment (mixmix) [inline method/group/add-pobox.js:21]: `tangles.group` gets auto-populated
`tangles.poBox` doesn't....

File: spec/lib/field-types.js [hunk 14]
```diff
@@ -1,6 +1,23 @@
...
 module.exports = {
-  string: overwrite({ type: ['string', 'null'] })
+  string: overwrite({ type: ['string', 'null'] }),
+  poBox: overwrite({
+    type: ['object', 'null'],
+    required: ['publicKey', 'secretKey'],
+    properties: {
+      publicKey: {
+        type: ['string'],
+        pattern: '^ssb://'
```
Comment (mixmix) [inline spec/lib/field-types.js:14]: change to `poBoxId`

File: test/spec/group/add-pobox/is-valid.test.js [hunk 15]
```diff
@@ -0,0 +1,71 @@
+const test = require('tape')
+const Crut = require('ssb-crut')
+const crypto = require('crypto')
+
```
Comment (mixmix) [inline test/spec/group/add-pobox/is-valid.test.js:4]: RM file

**Pull request merged**